File: backend/app/schemas/requirement.py
import uuid
from datetime import datetime
from typing import Dict, List, Optional
from pydantic import BaseModel, ConfigDict, Field


class RequirementEvidenceResponse(BaseModel):
    """Verifiable source evidence item supporting a requirement."""
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    requirement_id: uuid.UUID
    project_id: uuid.UUID
    artifact_id: Optional[uuid.UUID] = None
    source_type: str
    artifact_name: Optional[str] = None
    page_number: Optional[int] = None
    section_title: Optional[str] = None
    exact_snippet: str
    extraction_method: str
    confidence: float = 1.0
    created_at: datetime


class RequirementResponse(BaseModel):
    """Structured requirement entity."""
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    project_id: uuid.UUID
    requirement_id: str
    title: str
    description: str
    category: str
    priority: Optional[str] = None
    actor: Optional[str] = None
    status: str
    is_ambiguous: bool = False
    conflict_summary: Optional[str] = None
    content_hash: str
    evidence_count: int = 0
    created_at: datetime
    updated_at: datetime


class RequirementDetailResponse(RequirementResponse):
    """Detailed requirement view including all supporting evidence items."""
    evidence: List[RequirementEvidenceResponse] = Field(default_factory=list)


class RequirementExtractionSummaryResponse(BaseModel):
    """Summary of requirement extraction execution outcome."""
    total_extracted: int
    functional_count: int
    non_functional_count: int
    security_count: int
    ambiguous_count: int
    conflicted_count: int
    requirements: List[RequirementResponse]


class RequirementMetricsResponse(BaseModel):
    """Aggregate metrics and distribution for project requirements."""
    total: int
    by_category: Dict[str, int]
    by_priority: Dict[str, int]
    by_status: Dict[str, int]
    ambiguous_count: int
    conflict_count: int

File: frontend/src/types/requirement.ts
export interface RequirementEvidence {
  id: string;
  requirement_id: string;
  project_id: string;
  artifact_id: string | null;
  source_type: "artifact" | "project_metadata" | string;
  artifact_name: string | null;
  page_number: number | null;
  section_title: string | null;
  exact_snippet: string;
  extraction_method: string;
  confidence: number;
  created_at: string;
}

export interface Requirement {
  id: string;
  project_id: string;
  requirement_id: string;
  title: string;
  description: string;
  category: "functional" | "non_functional" | "security" | "performance" | "interface" | "general" | string;
  priority: string | null;
  actor: string | null;
  status: "extracted" | "conflicted" | "ambiguous" | "deprecated" | string;
  is_ambiguous: boolean;
  conflict_summary: string | null;
  content_hash: string;
  evidence_count: number;
  created_at: string;
  updated_at: string;
}

export interface RequirementDetail extends Requirement {
  evidence: RequirementEvidence[];
}

export interface RequirementExtractionSummary {
  total_extracted: number;
  functional_count: number;
  non_functional_count: number;
  security_count: number;
  ambiguous_count: number;
  conflicted_count: number;
  requirements: Requirement[];
}

export interface RequirementMetrics {
  total: number;
  by_category: Record<string, number>;
  by_priority: Record<string, number>;
  by_status: Record<string, number>;
  ambiguous_count: number;
  conflict_count: number;
}
